src/wp-includes/formatting.php

```diff
@@ -914,6 +917,169 @@ function seems_utf8( $str ) {
+ *
+ * Note that it’s unlikely for non-UTF-8 data to validate as UTF-8, but
+ * it is still possible. Many texts are simultaneously valid UTF-8,
+ * valid US-ASCII, and valid ISO-8859-1 (`latin1`).
+ *
+ * Example:
+ *
+ *     true === wp_is_valid_utf8( '' );
+ *     true === wp_is_valid_utf8( 'just a test' );
+ *     true === wp_is_valid_utf8( "\xE2\x9C\x8F" );    // Pencil, U+270F.
+ *     true === wp_is_valid_utf8( "\u{270F}" );        // Pencil, U+270F.
+ *     true === wp_is_valid_utf8( '✏' );              // Pencil, U+270F.
+ *
+ *     false === wp_is_valid_utf8( "just \xC0 test" ); // Invalid bytes.
+ *     false === wp_is_valid_utf8( "\xE2\x9C" );       // Invalid/incomplete sequences.
+ *     false === wp_is_valid_utf8( "\xC1\xBF" );       // Overlong sequences.
+ *     false === wp_is_valid_utf8( "\xED\xB0\x80" );   // Surrogate halves.
+ *     false === wp_is_valid_utf8( "B\xFCch" );        // ISO-8859-1 high-bytes.
+ *                                                     // E.g. The “ü” in ISO-8859-1 is a single byte 0xFC,
+ *                                                     // but in UTF-8 is the two-byte sequence 0xC3 0xBC.
+ *
+ * @see _wp_is_valid_utf8_fallback
+ *
+ * @since 6.9.0
+ *
+ * @param string $bytes String which might contain text encoded as UTF-8.
+ * @return bool Whether the provided bytes can decode as valid UTF-8.
+ */
+function wp_is_valid_utf8( string $bytes ): bool {
+	/*
+	 * Since PHP 8.3.0 the UTF-8 validity is cached internally
+	 * on string objects, making this a direct property lookup.
+	 *
+	 * This is to be preferred exclusively once PHP 8.3.0 is
+	 * the minimum supported version, because even when the
+	 * status isn’t cached, it uses highly-optimized code to
+	 * validate the byte stream.
+	 */
+	return function_exists( 'mb_check_encoding' )
+		? mb_check_encoding( $bytes, 'UTF-8' )
+		: _wp_is_valid_utf8_fallback( $bytes );
+}
+
+/**
+ * Fallback mechanism for safely validating UTF-8 bytes.
+ *
+ * By implementing a raw method here the code will behave
+ * in the same way on all installed systems, regardless of
+ * what extensions are installed.
+ *
+ * > [The following table] lists all of the byte sequences that
+ * > are well-formed in UTF-8. A range of byte values such as
+ * > [A0, BF] indicates that any byte from A0 to BF (inclusive)
+ * > is well-formed in that position. Any byte value outside
+ * > of the ranges listed is ill-formed.
+ *
+ *     First Byte    Second Byte   Third Byte    Fourth Byte
+ *     [0x00,0x7F]
+ *     [0xC2,0xDF]   [0x80,0xBF]
+ *     0xE0          [0xA0,0xBF]   [0x80,0xBF]
+ *     [0xE1,0xEC]   [0x80,0xBF]   [0x80,0xBF]
+ *     0xED          [0x80,0x9F]   [0x80,0xBF]
+ *     [0xEE,0xEF]   [0x80,0xBF]   [0x80,0xBF]
+ *     0xF0          [0x90,0xBF]   [0x80,0xBF]   [0x80,0xBF]
+ *     [0xF1,0xF3]   [0x80,0xBF]   [0x80,0xBF]   [0x80,0xBF]
+ *     0xF4          [0x80,0x8F]   [0x80,0xBF]   [0x80,0xBF]
```

Comment:
I find this table difficult to read, in I'd like a first column with codepoint range. [Could we go with something closer to the table used in the spec you reference?](https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-3/#G27506)

```suggestion
 *     Code Points          Byte 1       Byte 2       Byte 3       Byte 4
 *     U+0000..U+007F       0x00..0x7F
 *     U+0080..U+07FF       0xC2..0xDF   0x80..0xBF
 *     U+0800..U+0FFF       0xE0         0xA0..0xBF   0x80..0xBF
 *     U+1000..U+CFFF       0xE1..0xEC   0x80..0xBF   0x80..0xBF
 *     U+D000..U+D7FF       0xED         0x80..0x9F   0x80..0xBF
 *     U+E000..U+FFFF       0xEE..0xEF   0x80..0xBF   0x80..0xBF
 *     U+10000..U+3FFFF     0xF0         0x90..0xBF   0x80..0xBF   0x80..0xBF
 *     U+40000..U+FFFFF     0xF1..0xF3   0x80..0xBF   0x80..0xBF   0x80..0xBF
 *     U+100000..U+10FFFF   0xF4         0x80..0x8F   0x80..0xBF   0x80..0xBF
```

Reply:
converted the table as requested in e0fb4dab5d6d9115dc2172645df32b6775449a32